lib/Dialect/FIRRTL/Transforms/InferResets.cpp

```diff
@@ -1400,6 +1400,46 @@ LogicalResult InferResetsPass::implementAsyncReset(FModuleOp module,
       opsToUpdate.push_back(op);
   });
 
+  // If the reset is a local wire or node, move it upwards such that it
+  // dominates all the operations that it will need to attach to. In the case of
+  // a node this might not be easily possible, so we just spill into a wire in
+  // that case.
+  if (!actualReset.isa<BlockArgument>()) {
+    // The first op in `opsToUpdate` is the first op in the block, since `walk`
+    // is performed in block order. So we can simply check if the local reset
+    // declaration is before the first op to find out if we need to move
+    // anything.
+    auto *resetOp = actualReset.getDefiningOp();
+    if (!opsToUpdate.empty() && !resetOp->isBeforeInBlock(opsToUpdate[0])) {
```

Comment:
Is there any potential problem using `isBeforeInBlock` when the `opToUpdate` is inside a WhenOp, or some other reason the ops wouldn't be in the same block?

Reply:
Yeah that is an excellent point. That is likely to throw things off here. Looks like this needs to be more careful about finding the common ancestor block between the ops to actually figure out where things should get inserted.

Reply:
Added additional tests and updated the code to work with operations in different nested blocks.